src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -438,40 +438,43 @@ private bool AllFixed()
             return true;
         }
 
-        private void AddLowerBound(TypeParameterSymbol methodTypeParameter, TypeSymbol lowerBound)
+        private void AddBound(TypeSymbol addedBound, Dictionary<TypeSymbol, TypeSymbol>[] collectedBounds, TypeParameterSymbol methodTypeParameter)
         {
             Debug.Assert(IsUnfixedTypeParameter(methodTypeParameter));
             int methodTypeParameterIndex = methodTypeParameter.Ordinal;
-            if (_lowerBounds[methodTypeParameterIndex] == null)
+
+            if (collectedBounds[methodTypeParameterIndex] == null)
             {
-                _lowerBounds[methodTypeParameterIndex] = new HashSet<TypeSymbol>();
+                collectedBounds[methodTypeParameterIndex] = new Dictionary<TypeSymbol, TypeSymbol>(EqualsIgnoringComparer.Instance);
             }
 
-            _lowerBounds[methodTypeParameterIndex].Add(lowerBound);
+            if (collectedBounds[methodTypeParameterIndex].TryGetValue(addedBound, out TypeSymbol found))
```

Comment:
Fetch "collectedBounds[methodTypeParameterIndex]" into a local. No need to dig into dictionary this often. 
Old code could benefit from that as well.